File: src/main/java/googology/core/Notation.java
package googology.core;

/** Generic notation base. Every notation extends this class. */
public abstract class Notation {
    // --- taxonomy ---
    public abstract String name();
    public abstract Family family();
    public String subfamily() { return ""; }
    public String style() { return ""; }

    // --- attribution ---
    public String creator() { return ""; }
    public String version() { return ""; }

    // --- capability ---
    public abstract Capabilities capabilities();
    public boolean can(Op op) { return capabilities().has(op); }

    // --- string conversion ---
    public void stringToIt(String s) {
        throw new UnsupportedOperation(name(), Op.FROM_STRING);
    }

    public abstract String toLatex();

    // --- operations ---
    public int compare(Notation other) {
        throw new NotComparable(name());
    }

    public Notation expand(int n) {
        throw new UnsupportedOperation(name(), Op.EXPAND);
    }

    public Notation expandTo(int length) {
        throw new UnsupportedOperation(name(), Op.EXPAND_TO);
    }

    public boolean comparable() { return can(Op.COMPARE); }

    public Notation reduce() {
        String prev = toLatex();
        for (int i = 0; i < 1000; i++) {
            expand(1);
            String cur = toLatex();
            if (cur.equals(prev)) return this;
            prev = cur;
        }
        return this;
    }

    @Override
    public String toString() { return toLatex(); }
}

File: src/GoogologyLib/Core/Notation.cs
namespace GoogologyLib.Core;

/// <summary>Exception for unsupported operations.</summary>
public class UnsupportedOperation : InvalidOperationException
{
    public UnsupportedOperation(string name, Op op)
        : base($"Notation '{name}' does not implement the requested operation") { }
}

/// <summary>Exception for non-comparable notations (large-number notations).</summary>
public class NotComparable : InvalidOperationException
{
    public NotComparable(string name)
        : base($"Notation '{name}': comparison is undefined for large-number notations") { }
}

/// <summary>Generic notation base. Every notation derives from this class.</summary>
public abstract class Notation
{
    // --- taxonomy ---
    public abstract string Name { get; }
    public abstract Family Family { get; }
    public virtual string Subfamily => "";
    public virtual string Style => "";

    // --- attribution ---
    public virtual string Creator => "";
    public virtual string Version => "";

    // --- capability ---
    public abstract Capabilities Capabilities { get; }
    public bool Can(Op op) => Capabilities.Has(op);

    // --- string conversion ---
    public virtual void StringToIt(string s) =>
        throw new UnsupportedOperation(Name, Op.FromString);

    public abstract string ToLatex();

    // --- operations ---
    public virtual int Compare(Notation other) =>
        throw new NotComparable(Name);

    public virtual Notation Expand(int n = 1) =>
        throw new UnsupportedOperation(Name, Op.Expand);

    public virtual Notation ExpandTo(int length) =>
        throw new UnsupportedOperation(Name, Op.ExpandTo);

    public bool Comparable => Can(Op.Compare);

    public Notation Reduce()
    {
        var prev = ToLatex();
        for (int i = 0; i < 1000; i++)
        {
            Expand(1);
            var cur = ToLatex();
            if (cur == prev) return this;
            prev = cur;
        }
        return this;
    }

    public override string ToString() => ToLatex();
}